packages/web3/src/codec/transaction-codec.ts

```diff
@@ -0,0 +1,110 @@
+/*
+Copyright 2018 - 2022 The Alephium Authors
+This file is part of the alephium project.
+
+The library is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+The library is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with the library. If not, see <http://www.gnu.org/licenses/>.
+*/
+import { Parser } from 'binary-parser'
+import { ArrayCodec, DecodedArray } from './array-codec'
+import { Codec } from './codec'
+import { UnsignedTransactionCodec, UnsignedTransaction, unsignedTransactionCodec } from './unsigned-transaction-codec'
+import { Signature, signatureCodec } from './signature-codec'
+import { ContractOutputRef, contractOutputRefCodec } from './contract-output-ref-codec'
+import { Either, EitherCodec } from './either-codec'
+import { AssetOutput, AssetOutputCodec, assetOutputCodec } from './asset-output-codec'
+import { ContractOutput, ContractOutputCodec, contractOutputCodec } from './contract-output-codec'
+import { Transaction as Tx } from '../api/api-alephium'
+import { hexToBinUnsafe } from '..'
+
+type Output = Either<AssetOutput, ContractOutput>
+const outputCodec = new EitherCodec<AssetOutput, ContractOutput>(assetOutputCodec, contractOutputCodec)
+const outputsCodec = new ArrayCodec(outputCodec)
+export const signaturesCodec = new ArrayCodec(signatureCodec)
+const contractOutputRefsCodec = new ArrayCodec(contractOutputRefCodec)
+
+export interface Transaction {
+  unsigned: UnsignedTransaction
+  scriptExecutionOk: number
+  contractInputs: DecodedArray<ContractOutputRef>
+  generatedOutputs: DecodedArray<Output>
+  inputSignatures: DecodedArray<Signature>
+  scriptSignatures: DecodedArray<Signature>
```

Comment:
We need to call the `convertTo*` function after decoding, is it possible to directly declare the type to `ContractOutputRef[]`, `Output[]`, etc?

Or we can add some util functions like this:

```
function decodeTransaction(bytes: Uint8Array): Transaction {
  const tx = transactionCodec.decode(bytes)
  return TransactionCodec.convertTo(tx)
}
```

Reply:
Sorry for the late reply, will do :+1: